tests/test_config_entries.py

```diff
@@ -2882,6 +2899,189 @@ async def async_step_user(self, user_input=None):
+            config_entries.SOURCE_IGNORE,
+            "mock-unique-id",
+            {"discovery_key": {"domain": "test", "key": "blah", "version": 1}},
+            config_entries.SOURCE_ZEROCONF,
+            data_entry_flow.FlowResultType.ABORT,
+            {"discovery_keys": [{"domain": "test", "key": "blah", "version": 1}]},
+        ),
+        # Discovery key not added to user entry data
+        (
+            {},
+            config_entries.SOURCE_USER,
+            "mock-unique-id",
+            {"discovery_key": {"domain": "test", "key": "blah", "version": 1}},
+            config_entries.SOURCE_ZEROCONF,
+            data_entry_flow.FlowResultType.ABORT,
+            {},
+        ),
+        # Flow not aborted when unique id is not matching
+        (
+            {},
+            config_entries.SOURCE_IGNORE,
+            "mock-unique-id-2",
+            {"discovery_key": {"domain": "test", "key": "blah", "version": 1}},
+            config_entries.SOURCE_ZEROCONF,
+            data_entry_flow.FlowResultType.FORM,
+            {},
+        ),
+        # Flow not aborted when user initiated flow
+        (
+            {},
+            config_entries.SOURCE_IGNORE,
+            "mock-unique-id-2",
+            {"discovery_key": {"domain": "test", "key": "blah", "version": 1}},
+            config_entries.SOURCE_USER,
+            data_entry_flow.FlowResultType.FORM,
+            {},
+        ),
+    ],
+)
+async def test_ignored_entry_update_discovery_keys(
+    hass: HomeAssistant,
+    manager: config_entries.ConfigEntries,
+    entry_data: dict,
+    entry_source: str,
+    entry_unique_id: str,
+    flow_context: dict,
+    flow_source: str,
+    flow_result: data_entry_flow.FlowResultType,
+    updated_entry_data: dict,
+) -> None:
+    """Test that discovery keys of an ignored entry can be updated."""
+    hass.config.components.add("comp")
+    entry = MockConfigEntry(
+        domain="comp",
+        data=entry_data,
+        unique_id=entry_unique_id,
+        state=config_entries.ConfigEntryState.LOADED,
+        source=entry_source,
+    )
+    entry.add_to_hass(hass)
+
+    mock_integration(hass, MockModule("comp"))
+    mock_platform(hass, "comp.config_flow", None)
+
+    class TestFlow(config_entries.ConfigFlow):
+        """Test flow."""
+
+        VERSION = 1
+
+        async def async_step_user(self, user_input=None):
+            """Test user step."""
+            await self.async_set_unique_id("mock-unique-id")
+            self._abort_if_unique_id_configured(reload_on_update=False)
+            return self.async_show_form(step_id="step2")
+
+        async def async_step_step2(self, user_input=None):
+            raise NotImplementedError
+
+        async def async_step_zeroconf(self, discovery_info=None):
+            """Test zeroconf step."""
+            return await self.async_step_user(discovery_info)
+
+    with (
+        patch.dict(config_entries.HANDLERS, {"comp": TestFlow}),
```

Comment:
Should we use the fixture that does this instead?

Reply:
yeah, definitely. Tests added in the PR should use the fixture, and existing tests improved in a separate PR.